rfc/0019-reflectors-database-cache.md

```diff
@@ -0,0 +1,164 @@
+It stores the watched objects in memory, and it is responsible for keeping them up-to-date.
+It also offers a reader interface to query the objects stored in the cache.
+At the time of writing, the reflectors are used to speed up the evaluation of policies that require information about the Kubernetes cluster state,
+by avoiding fetching the same objects from the Kubernetes API server multiple times during the evaluation of a single request.
+However, the reflectors cache is currently stored in memory, which can lead to high memory usage in the policy server.
+According to the [kube-rs optimizian guide](https://kube.rs/controllers/optimization/#reflector-optimization) and the [reflector documentation](https://docs.rs/kube/latest/kube/runtime/reflector/fn.reflector.html#memory-usage), the cache [Store](https://docs.rs/kube-runtime/0.88.1/src/kube_runtime/reflector/store.rs.html#13) is the main source of memory usage and it is directly proportional to how many objects are stored.
+In the worst case scenario (Pods with many sidecar containers and many environment variables), the reflectors cache can consume up to 1GB of memory for ~2000 objects.
+It's possible to reduce the memory usage by dropping objects managed fields, however this is still not enough in large clusters.
+Also, every time a `fieldSelector` or a `labelSelector` is used, a new reflector is created, which increases the memory usage by possibly duplicating the objects in memory.
+At the time of writing, reflectors are created in a lazy manner the first time they are needed.
+
+The current flow is as follows:
+
+1. The policy server receives a request to evaluate a policy.
+2. The policy server creates a new reflector for each resource kind that is being requested by the policy.
+3. The reflector fetches the objects from the Kubernetes API server and stores them in memory, this step is triggered by the event `watcher::Event::Restarted`.
+4. The policy server evaluates the policy using the objects stored in the reflectors cache.
+5. The reflector keeps the objects up-to-date by watching for changes in the Kubernetes API server.
+
+## SQLite database store
+
+This RFC proposes to create a custom `Store` that uses a disk-based database to store the objects.
+The database will be used to store the objects in JSON format, and it will be queried using the `fieldSelector` and `labelSelector` to reduce the number of reflectors that are created and maintained in memory.
+SQLite is a good candidate for the database, as it is a single file database and supports SQL and JSON queries.
+Also, the database can be considered ephemeral, as the reflectors can be recreated from the Kubernetes API server if needed.
+The database will contain a table for each resource kind that is being watched by the reflectors.
+
+Table schema:
```

Comment:
You need to mention we're going to create one table per type of watched request. Plus explain the naming convention we want to use to name these tables.

Finally, point out that by having 1 table per resource we reduce the concurrency issues when inserting/querying data.